pkgs/sketch_pad/lib/editor/editor.dart

```diff
@@ -216,45 +220,52 @@ class _EditorWidgetState extends State<EditorWidget> implements EditorService {
     codeMirror?.setTheme(darkMode ? 'monokai' : 'default');
   }
 
-  Future<HintResults> _calculateCompletions() async {
-    final editor = codeMirror!;
-    final doc = editor.doc;
-    final offset = doc.indexFromPos(doc.getCursor()) ?? 0;
+  Future<HintResults> _completions(CodeMirror _, [HintsOptions? __]) async {
+    final operation = completionType;
+    completionType = CompletionType.auto;
 
     final appServices = widget.appServices;
-    final response = await appServices.services
-        .complete(services.SourceRequest(
-          source: doc.getValue() ?? '',
-          offset: offset,
-        ))
-        .onError(
-          (error, st) => services.CompleteResponse(
-              replacementLength: 0, replacementOffset: 0, suggestions: []),
-        );
-
-    final replaceOffset = response.replacementOffset;
-    final replaceLength = response.replacementLength;
-
-    final completions = response.suggestions.map((suggestion) =>
-        AnalysisCompletion(replaceOffset, replaceLength, suggestion));
-    final hints =
-        completions.map((completion) => completion.toCodemirrorHint()).toList();
-
-    // Remove hints where both the replacement text and the display text is the
-    // same.
-    final memos = <String>{};
-    hints.retainWhere((hint) {
-      final memo = '${hint.text}:${hint.displayText}';
-      if (memos.contains(memo)) return false;
-
-      memos.add(memo);
-      return true;
-    });
 
-    final from = doc.posFromIndex(replaceOffset);
-    final to = doc.posFromIndex(replaceOffset + replaceLength);
-
-    return HintResults.fromHints(hints, from, to);
+    final editor = codeMirror!;
+    final doc = editor.doc;
+    final source = doc.getValue() ?? '';
+    final sourceOffset = doc.indexFromPos(doc.getCursor()) ?? 0;
+
+    if (operation == CompletionType.quickfix) {
+      final response = await appServices.services
+          .fixes(services.SourceRequest(source: source, offset: sourceOffset))
+          .onError((error, st) => services.FixesResponse.empty);
+
+      if (response.fixes.isEmpty && response.assists.isEmpty) {
+        widget.appModel.editorStatus.showToast('No quick fixes available.');
+      }
+
+      return HintResults.fromHints([
+        ...response.fixes.map((change) => change.toHintResult()),
+        ...response.assists.map((change) => change.toHintResult()),
+      ], doc.posFromIndex(sourceOffset), doc.posFromIndex(0));
+    } else {
+      final response = await appServices.services
+          .complete(
+              services.SourceRequest(source: source, offset: sourceOffset))
+          .onError((error, st) => services.CompleteResponse.empty);
+
+      final offset = response.replacementOffset;
+      final length = response.replacementLength;
+      final hints = response.suggestions
+          .map((suggestion) => suggestion.toHintResult())
+          .toList();
+
+      // Remove hints where both the replacement text and the display text are the
+      // same.
```

Comment:
I really appreciate this comment, otherwise the next snippet would have made me get a bit lost.

Reply:
Likely we can do better here if we up-level the code completion in DartPad a bit. That would likely mean creating our own DOM based code completion, or a flutter web UI, or switching the editor out to monaco and using what they provide.